### PR TITLE
gps: avoid pow() in GPS_HARDSLEEP threshold heuristic

### PR DESCRIPTION
**Pull request by ndoo**
## What

`GPS::down()` picks between `GPS_SOFTSLEEP` and `GPS_HARDSLEEP` using `2750 * pow(seconds, 1.22)` - a curve fit the surrounding comment already describes as "not particularly accurate". Replaces it with `gpsHardsleepThresholdMs()`, a piecewise-linear lookup over 16 sampled points from the same curve, avoiding `pow()` in this one call site.

Verified via [CodeSizer](https://github.com/Wren6991/CodeSizer) (a call-graph-based static size profiler) that on `wio-e5`, this was the **only** `pow()` call site in the entire firmware - it single-handedly pulled in the full double-precision `libm` `pow`/`rem_pio2` chain for a threshold decision the code already admits is approximate.

## Accuracy of the replacement

Piecewise-linear interpolation between the 16 sampled points tracks the original formula within:
- ~0.5% for inputs >=10s
- ~1.6% for 5-10s
- larger relative (but small absolute - at most a couple of seconds) error below 5s

All negligible next to `gps_update_interval` values, which range from tens of seconds to hours. New native test suite (`test_gps_update_scheduling`) checks this against the original formula at sampled points, monotonicity, and the intentional clamp above the pre-existing 15-minute search-duration cap (`GPSUpdateScheduling::searchedTooLong`).

## Flash-size impact, measured across 4 boards

| Board | Chip | Result |
| --- | --- | --- |
| `wio-e5` | STM32WLE5 | **-3,104 bytes** (233,744 -> 230,640 B used, 94.3% -> 93.1% of the 247,808 B budget) |
| `seeed_wio_tracker_L1` | nRF52840 | ~0 net effect (+144 B measured, within noise/table overhead) |
| `heltec-v3` / `heltec-v4` | ESP32-S3 | Not reliably measurable - see note below |

`wio-e5` is the only board of the four where this was the sole `pow()` call site - confirmed via disassembly (`objdump`/`addr2line`) that `__ieee754_pow` had exactly one caller before this change. On `seeed_wio_tracker_L1`, `pow()` has **6 other call sites** that keep it linked regardless of this change: `GeoCoord::latLongToUTM`, `GeoCoord::ensureOSGR` (both reachable from `UIRenderer.cpp`'s on-screen coordinate display), and three sensor telemetry drivers (`OPT3001Sensor`, `MLX90632Sensor` x2, `BMP3XXSingleton`). `heltec-v3`/`heltec-v4` both build with `HAS_SCREEN=1` and general sensor-telemetry support, so the same mechanism almost certainly applies there too - but I couldn't get a trustworthy delta for them: rebuilding the *unmodified* `heltec-v3` baseline twice in a row (no code changes between runs) produced a ~20KB difference in final `.text`+`.data` size, which I traced to non-determinism in this repo's ESP-IDF/Xtensa LTO build rather than anything code-related. That noise floor is larger than any plausible effect of this change, so I'm not reporting a specific number for the ESP32-S3 boards rather than presenting a number I don't trust.

tl;dr: this specifically helps flash-constrained, GPS-enabled, screen-less/sensor-less boards like `wio-e5` (currently at 94%+ flash usage); it's a no-op-to-negligible on boards that already need `pow()` for other reasons.

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [ ] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [ ] Heltec (Lora32) V3
  - [ ] LilyGo T-Deck
  - [ ] LilyGo T-Beam
  - [ ] RAK WisBlock 4631
  - [ ] Seeed Studio T-1000E tracker card
  - [ ] Other (please specify below)

Tested via native unit tests (new `test_gps_update_scheduling` suite) and flash-size measurement across four boards, above. Not yet flashed to real hardware - this only changes an internal GPS sleep-mode threshold heuristic, not GPS communication itself, so I don't expect a behavioral regression, but flagging per the checklist above since I haven't verified on physical devices.

## Summary by CodeRabbit

* **Improvements**
  * Refined GPS sleep scheduling with an interpolated threshold based on predicted search duration.
  * Improved handling of zero-duration estimates and durations beyond the supported range.
  * Enables more consistent selection between soft sleep and hard sleep, helping balance responsiveness and power consumption.

* **Tests**
  * Added automated coverage for threshold accuracy, monotonic behavior, boundary values, and clamping.

### CODE REVIEW

**Comment by ndoo** [comment]
@coderabbitai review

**Comment by ndoo** [comment]
Trimmed the flagged comments to 1-2 lines each in a supplementary commit (the hardware-validation nitpick is already tracked in the Attestations checklist above - no physical device to test against here).

@coderabbitai review

**Comment by ndoo** [comment]
this seems low-impact enough to take out of draft, perhaps @todd-herbert would be the appropriate reviewer, looking at git blame…

**Review by caveman99 — APPROVED** [review]

**Comment by caveman99** [comment]
@ndoo please review https://github.com/meshmy/meshtastic-firmware/pull/8

**Comment by ndoo** [comment]
> @ndoo please review [meshmy#8](https://github.com/meshmy/meshtastic-firmware/pull/8)

Done

File: test/test_gps_update_scheduling/test_main.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+static void test_exact_at_table_breakpoints(void)
+{
+    // Every breakpoint must return its own sampled value. Catches an off-by-one in the segment
+    // scan, which a percentage bound on interpolated points would absorb.
+    const uint32_t breakpoints[] = {0, 1, 2, 3, 5, 10, 15, 20, 30, 45, 60, 90, 120, 180, 240, 300, 450, 600, 900};
+    for (uint32_t s : breakpoints) {
+        char msg[64];
+        snprintf(msg, sizeof(msg), "breakpoint %us", s);
```
[the remark this answers is not in the document]
Comment (ndoo) [reply test/test_gps_update_scheduling/test_main.cpp:56]: @coderabbitai I think this may be a non-issue on Arduino framework?
